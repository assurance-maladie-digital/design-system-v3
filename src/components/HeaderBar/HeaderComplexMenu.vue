<script setup lang="ts">
	import { mdiClose, mdiMenu } from '@mdi/js'
	import { computed, onMounted, onUnmounted, provide, readonly, ref, watch } from 'vue'
	import { useTheme, useDisplay } from 'vuetify'

	const theme = useTheme()
	const display = useDisplay()

	const menuBtnWrapper = ref<HTMLDivElement | null>(null)
	const menuLeft = ref(0)
	const menuTop = ref(0)

	function positionMenu() {
		// todo debounce
		menuLeft.value = menuBtnWrapper.value!.getBoundingClientRect().left
		menuTop.value = menuBtnWrapper.value!.getBoundingClientRect().top
	}

	onMounted(() => {
		positionMenu()
		window.addEventListener('scroll', positionMenu)
		window.addEventListener('resize', positionMenu)
	})

	onUnmounted(() => {
		window.removeEventListener('scroll', positionMenu)
		window.removeEventListener('resize', positionMenu)
	})

	const menuOpen = ref(false)

	function toggleMenu() {
		menuOpen.value = !menuOpen.value
	}

	watch(menuOpen, (newVal) => {
		document.documentElement.style.overflow = newVal ? 'hidden' : 'auto'
		document.body.style.overflow = newVal ? 'hidden' : 'auto'
	})

	const btnStyle = computed(() => ({
		backgroundColor: menuOpen.value ? '#fff' : theme.current.value.colors.primary,
		color: menuOpen.value ? theme.current.value.colors.primary : '#fff',
	}))

	const menuStyle = computed(() => ({
		left: `${menuLeft.value}px`,
		top: `${menuTop.value}px`,
	}))

	function handleClickOutside(event: MouseEvent | KeyboardEvent) {
		if (!menuOpen.value) return

		// do not close menu if click is inside the menu
		const menu = document.getElementById('header-menu-wrapper')
		let walkElement = event.target as HTMLElement | null
		while (walkElement && walkElement !== document.body) {
			if (walkElement === menu) return
			walkElement = walkElement.parentElement
		}

		menuOpen.value = false
	}

	const childMenuOpen = ref(false)
	function defineChildMenuOpen(value: boolean) {
		childMenuOpen.value = value
	}
	provide('parentMenuOpen', readonly(menuOpen))
	provide('defineChildMenuOpen', defineChildMenuOpen)
</script>

<template>
	<div>
		<div ref="menuBtnWrapper">
			<button
				class="header-menu-btn mr-4"
				:style="btnStyle"
				@click="toggleMenu"
			>
				<VIcon size="48">
					{{ menuOpen ? mdiClose : mdiMenu }}
				</VIcon>
				<span v-if="display.mdAndUp.value">Menu</span>
			</button>
		</div>
		<div
			v-show="menuOpen"
			class="menu overlay"
			role="menu"
			tabindex="0"
			@click="handleClickOutside"
			@keydown.enter="handleClickOutside"
			@keydown.esc="handleClickOutside"
		>
			<div
				class="menu-wrapper"
				:style="menuStyle"
			>
				<button
					class="header-menu-btn mr-4"
					:style="btnStyle"
					@click="toggleMenu"
				>
					<VIcon size="48">
						{{ menuOpen ? mdiClose : mdiMenu }}
					</VIcon>
					<span v-if="display.mdAndUp.value">Menu</span>
				</button>
				<nav
					id="header-menu-wrapper"
					class="header-menu-wrapper"
					:class="{
						'header-menu-wrapper--submenu-open': childMenuOpen,
					}"
				>
					<div class="header-menu">
						<slot />
					</div>
				</nav>
			</div>
		</div>
	</div>
</template>

<style lang="scss" scoped>
$header-height: 82px;

.overlay {
	inset: 0;
	position: fixed;
	z-index: 1000;
	background-color: rgba(3, 16, 37, .5);
}

.header-menu-btn {
	text-transform: Capitalize;
	height: 82px;
	width: 77px;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
}

.menu-wrapper {
	height: 100dvh;
	background-color: #fff;
	display: flex;
	flex-direction: column;
}

.header-menu-wrapper {
	height: calc(100% - $header-height);
	display: grid;
	position: relative;
	overflow: auto;
}

.header-menu-wrapper--submenu-open {
	overflow: clip;
}

@media screen and (min-width: 900px) {
	.header-menu-btn {
		height: 95px;
		width: 95px;
	}

	.menu-wrapper {
		position: absolute;
		background-color: transparent;
	}

	.header-menu-wrapper {
		height: 70vh;
		width: 350px;
		overflow: visible;
	}

	.header-menu {
		background-color: #fff;
		overflow-y : auto;
		overflow-x: hidden;
		height: 70vh;
	}
}
</style>
