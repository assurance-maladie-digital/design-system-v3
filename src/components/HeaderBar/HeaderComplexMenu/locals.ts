export default {
	mainMenu: 'Menu principal',
	publicMenu: 'Menu public',
}
