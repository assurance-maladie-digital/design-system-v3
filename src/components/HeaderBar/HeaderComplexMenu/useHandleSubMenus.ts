import { computed, provide, ref, watch, type DeepReadonly, type Ref } from 'vue'
import { registerSubMenuKey } from './conts'

export default function useHandleSubMenus(openStatus: DeepReadonly<Ref<boolean>>) {
	type SubMenu = { id: string, status: Ref<boolean>, close: () => void }
	const subMenus: Ref<SubMenu[]> = ref([])

	function registerSubMenu(status: Ref<boolean>, close: () => void) {
		const id = String(subMenus.value.length)
		const newSubMenu = { id, status, close }

		// Register the new submenu
		subMenus.value.push(newSubMenu)

		// Watch for changes to the status of this specific submenu
		watch(status, (newStatus) => {
			if (newStatus) {
				closeOtherSubMenus(newSubMenu)
			}
		})
	}

	function closeOtherSubMenus(subMenu: SubMenu) {
		subMenus.value.forEach((otherSubMenu) => {
			if (otherSubMenu.id !== subMenu.id && otherSubMenu.status) {
				otherSubMenu.close()
			}
		})
	}

	// When the current menu is closed, close its submenus
	watch(openStatus, (newOpenStatus) => {
		if (!newOpenStatus) {
			subMenus.value.forEach((subMenu) => {
				if (subMenu.status) {
					subMenu.close()
				}
			})
		}
	})

	const haveOpenSubMenu = computed(() => subMenus.value.some(subMenu => subMenu.status))

	provide(registerSubMenuKey, registerSubMenu)

	return {
		haveOpenSubMenu,
	}
}
