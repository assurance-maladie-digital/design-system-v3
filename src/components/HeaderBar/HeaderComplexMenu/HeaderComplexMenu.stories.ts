import HeaderMenu from './HeaderComplexMenu.vue'
import HeaderBar from '../HeaderBar.vue'
import type { Meta, StoryObj } from '@storybook/vue3'
import HeaderMenuSection from './HeaderMenuSection/HeaderMenuSection.vue'
import HeaderSubMenu from './HeaderSubMenu/HeaderSubMenu.vue'
import HeaderMenuItem from './HeaderMenuItem/HeaderMenuItem.vue'
import { VBtn } from 'vuetify/components'

const meta = {
	title: 'Components/HeaderBar/HeaderComplexMenu',
	component: HeaderMenu,
	parameters: {
		layout: 'fullscreen',
	},
} satisfies Meta<typeof HeaderMenu>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {},
	render: (args) => {
		return {
			components: { HeaderMenu, HeaderBar },
			setup() {
				return { args }
			},
			template: `
				<HeaderBar>
					<template #menu>
						<HeaderMenu>
							<p>lorem ipsum</p>
						</HeaderMenu>
					</template>
				</HeaderBar>
			`,
		}
	},
}

export const Populated: Story = {
	args: {},
	render: (args) => {
		return {
			components: { HeaderMenuItem, HeaderMenu, HeaderBar, HeaderSubMenu, HeaderMenuSection, VBtn },
			setup() {
				return { args }
			},
			template: `
				<HeaderBar>
					<template #menu>
						<HeaderMenu>
							<HeaderMenuSection title="section 1">
								<HeaderMenuItem>
									<a>Item 1</a>
								</HeaderMenuItem>
								<HeaderMenuItem>
									<a>Item 2</a>
								</HeaderMenuItem>
								<headerMenuItem>
									<HeaderSubMenu>
										<template #title>
											Menu de premier niveau 1
										</template>
										<HeaderMenuSection title="Section">
											<HeaderMenuItem>
												<a>Item</a>
											</HeaderMenuItem>
											<HeaderSubMenu>
												<template #title>
													Menu de deuxième niveau 1
												</template>
												<HeaderMenuItem>
													<a>Item</a>
												</HeaderMenuItem>
											</HeaderSubMenu>
										</HeaderMenuSection>
									</HeaderSubMenu>
								</headerMenuItem>
							</HeaderMenuSection>
							<HeaderMenuSection title="section 2">
								<headerMenuItem>
									<HeaderSubMenu>
										<template #title>
											Menu de premier niveau 2
										</template>
										<HeaderMenuItem>
											<a>Item 1</a>
										</HeaderMenuItem>
										<HeaderMenuItem>
											<HeaderSubMenu>
												<template #title>
													Menu de deuxième niveau 2
												</template>
												<HeaderMenuItem>
													<a>Item 1</a>
												</HeaderMenuItem>
											</HeaderSubMenu>
										</HeaderMenuItem>
										<HeaderMenuItem>
											<HeaderSubMenu>
												<template #title>
													Menu de deuxième niveau 3
												</template>
												<HeaderMenuSection title="section 1">
													<HeaderMenuItem>
														<a>Item 1</a>
													</HeaderMenuItem>
												</HeaderMenuSection>
											</HeaderSubMenu>
										</HeaderMenuItem>
									</HeaderSubMenu>
								</headerMenuItem>
								<HeaderMenuItem>
									<a>Item 3</a>
								</HeaderMenuItem>
							</HeaderMenuSection>
							<div class="pa-4">
								<p class="font-weight-bold">Veillez vous connecter</p>
								<VBtn variant="tonal" class="mt-4 font-weight-medium" color="primary">Je me connecte</VBtn>
							</div>
						</HeaderMenu>
					</template>
				</HeaderBar>
			`,
		}
	},
}
