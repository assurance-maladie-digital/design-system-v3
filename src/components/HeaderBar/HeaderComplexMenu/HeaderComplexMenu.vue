<script setup lang="ts">
	import { computed, inject, onMounted, onUnmounted, readonly, ref, watch, type Ref } from 'vue'
	import HeaderMenuBtn from '../HeaderMenuBtn/HeaderMenuBtn.vue'
	import useHandleSubMenus from './useHandleSubMenus'
	import locals from './locals'

	const headerMenuWrapper = ref<HTMLElement | null>(null)
	const menuBtnWrapper = ref<HTMLDivElement | null>(null)
	const menuLeft = ref(0)
	const menuTop = ref(0)

	function positionMenu() {
		// todo debounce
		menuLeft.value = menuBtnWrapper.value!.getBoundingClientRect().left
		menuTop.value = menuBtnWrapper.value!.getBoundingClientRect().top
	}

	onMounted(() => {
		positionMenu()
		window.addEventListener('scroll', positionMenu)
		window.addEventListener('resize', positionMenu)
	})

	onUnmounted(() => {
		window.removeEventListener('scroll', positionMenu)
		window.removeEventListener('resize', positionMenu)
	})

	const menuOpen = ref(false)

	watch(menuOpen, (newVal) => {
		document.documentElement.style.overflow = newVal ? 'hidden' : 'auto'
		document.body.style.overflow = newVal ? 'hidden' : 'auto'
	})

	const menuStyle = computed(() => ({
		left: `${menuLeft.value}px`,
		top: `${menuTop.value}px`,
	}))

	function handleClickOutside(event: MouseEvent | KeyboardEvent) {
		if (!menuOpen.value) return

		// do not close menu if click is inside the menu
		let walkElement = event.target as HTMLElement | null
		while (walkElement && walkElement !== document.body) {
			if (walkElement === headerMenuWrapper.value) return
			walkElement = walkElement.parentElement
		}

		menuOpen.value = false
	}

	const { haveOpenSubMenu } = useHandleSubMenus(readonly(menuOpen))

	const registerHeaderMenu = inject<(menuOpen: Ref<boolean>) => void>('registerHeaderMenu')
	if (registerHeaderMenu) registerHeaderMenu(menuOpen)
</script>

<template>
	<div
		role="dialog"
		aria-modal="true"
		:aria-label="locals.mainMenu"
	>
		<div ref="menuBtnWrapper">
			<HeaderMenuBtn v-model="menuOpen" />
		</div>
		<div
			v-show="menuOpen"
			class="menu overlay"
			role="menu"
			tabindex="0"
			@click="handleClickOutside"
			@keydown.enter="handleClickOutside"
			@keydown.esc="handleClickOutside"
		>
			<div
				class="menu-wrapper"
				:style="menuStyle"
			>
				<HeaderMenuBtn v-model="menuOpen" />
				<nav
					id="header-menu-wrapper"
					ref="headerMenuWrapper"
					class="header-menu-wrapper"
					:class="{
						'header-menu-wrapper--submenu-open': haveOpenSubMenu,
					}"
					role="navigation"
					:aria-label="locals.publicMenu"
				>
					<div class="header-menu">
						<slot />
					</div>
				</nav>
			</div>
		</div>
	</div>
</template>

<style lang="scss" scoped>
@use '@/assets/tokens.scss' as *;
@use '../consts' as *;

.overlay {
	inset: 0;
	position: fixed;
	z-index: 1000;
	background-color: rgba(3, 16, 37, .5);
}

.menu-wrapper {
	height: 100dvh;
	background-color: $neutral-white;
	display: flex;
	flex-direction: column;
}

.header-menu-wrapper {
	height: calc(100% - $header-height);
	display: grid;
	position: relative;
	overflow: auto;
}

.header-menu-wrapper--submenu-open {
	overflow: clip;
}

@media screen and (min-width: $header-breakpoint) {
	.menu-wrapper {
		position: absolute;
		background-color: transparent;
	}

	.header-menu-wrapper {
		height: $menu-height;
		width: $menu-width;
		overflow: visible;
	}

	.header-menu {
		background-color: $neutral-white;
		overflow-y : auto;
		overflow-x: hidden;
		height: $menu-height;
	}
}
</style>
