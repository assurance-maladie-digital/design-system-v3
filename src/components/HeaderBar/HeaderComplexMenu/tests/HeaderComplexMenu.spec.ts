import { vuetify } from '@tests/unit/setup'
import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import HeaderComplexMenu from '../HeaderComplexMenu.vue'
import { afterEach } from 'node:test'

describe('HeaderComplexMenu', () => {
	const BtnTestComponent = {
		setup() {
			const props = defineProps({ modelValue: Boolean })
			return { open: props.modelValue }
		},
		template: `<button @click="$emit('update:modelValue', !open)">Test</button>`,
	}

	afterEach(() => {
		vi.resetAllMocks()
		document.body.innerHTML = ''
	})

	it('should render the component', async () => {
		const wrapper = mount(HeaderComplexMenu, {
			global: {
				plugins: [vuetify],
				provide: {
					registerHeaderMenu: () => {},
				},
			},
			slots: {
				default: '<div>Default slot</div>',
			},
			stubs: {
				HeaderMenuBtn: BtnTestComponent,
			},
		})

		expect(wrapper.html()).toMatchSnapshot()

		const menu = wrapper.find('.overlay')
		expect(menu.attributes('style')).toContain('display: none;')

		const btn = wrapper.find('.header-menu-btn')
		await btn.trigger('click')
		expect(menu.attributes('style')).toBeUndefined()
	})

	it('should close the menu when clicking outside', async () => {
		const wrapper = mount(HeaderComplexMenu, {
			global: {
				plugins: [vuetify],
				provide: {
					registerHeaderMenu: () => {},
				},
			},
			stubs: {
				HeaderMenuBtn: BtnTestComponent,
			},
			slots: {
				default: '<div>Default slot</div>',
			},
			attachTo: document.body,
		})

		const overlay = wrapper.find('.overlay')
		const btn = wrapper.find('.header-menu-btn')

		await btn.trigger('click')
		expect(overlay.attributes('style')).toBeUndefined()

		await overlay.trigger('click')
		expect(overlay.attributes('style')).toContain('display: none;')

		wrapper.unmount()
	})

	it('should not close the menu when clicking inside', async () => {
		const wrapper = mount(HeaderComplexMenu, {
			global: {
				plugins: [vuetify],
				provide: {
					registerHeaderMenu: () => {},
				},
			},
			stubs: {
				HeaderMenuBtn: BtnTestComponent,
			},
			slots: {
				default: '<div>Default slot</div>',
			},
			attachTo: document.body,
		})

		const menu = wrapper.find('.overlay')
		const btn = wrapper.find('button')
		await btn.trigger('click')
		expect(menu.attributes('style')).toBeUndefined()

		await wrapper.find('.header-menu').trigger('click')
		expect(menu.attributes('style')).toBeUndefined()

		wrapper.unmount()
	})

	it('should listen to the button to open and close the menu', async () => {
		const wrapper = mount(HeaderComplexMenu, {
			global: {
				plugins: [vuetify],
				provide: {
					registerHeaderMenu: () => {},
				},
			},
			stubs: {
				HeaderMenuBtn: BtnTestComponent,
			},
			slots: {
				default: '<div>Default slot</div>',
			},
		})

		const menu = wrapper.find('.overlay')
		const btn = wrapper.find('button')
		await btn.trigger('click')
		expect(menu.attributes('style')).toBeUndefined()

		await btn.trigger('click')
		expect(menu.attributes('style')).toContain('display: none;')
	})
})
