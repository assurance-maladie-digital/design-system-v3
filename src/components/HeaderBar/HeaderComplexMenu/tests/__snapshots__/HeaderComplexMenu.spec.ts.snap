// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderComplexMenu > should render the component 1`] = `
"<div data-v-10d9bc74="" role="dialog" aria-modal="true" aria-label="Menu principal">
  <div data-v-10d9bc74=""><button data-v-70557900="" data-v-10d9bc74="" class="header-menu-btn" style="background-color: #1867C0; color: #fff;" type="button" aria-label="Ouvrir le menu" title="Ouvrir le menu"><i data-v-70557900="" class="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z mdi v-icon notranslate v-theme--light" style="font-size: 48px; height: 48px; width: 48px;" aria-hidden="true"></i><span data-v-70557900="" class="header-menu-btn__label">Menu</span></button></div>
  <transition-stub data-v-10d9bc74="" name="menu" appear="false" persisted="true" css="true">
    <div data-v-10d9bc74="" class="overlay" style="display: none;">
      <div data-v-10d9bc74="" role="menu" class="menu-wrapper" style="left: 0px; top: 0px;"><button data-v-70557900="" data-v-10d9bc74="" class="header-menu-btn" style="background-color: #1867C0; color: #fff;" type="button" aria-label="Ouvrir le menu" title="Ouvrir le menu"><i data-v-70557900="" class="M3,6H21V8H3V6M3,11H21V13H3V11M3,16H21V18H3V16Z mdi v-icon notranslate v-theme--light" style="font-size: 48px; height: 48px; width: 48px;" aria-hidden="true"></i><span data-v-70557900="" class="header-menu-btn__label">Menu</span></button>
        <nav data-v-10d9bc74="" id="header-menu-wrapper" class="header-menu-wrapper" role="navigation" aria-label="Menu public">
          <div data-v-10d9bc74="" class="header-menu">
            <div>Default slot</div>
          </div>
        </nav>
      </div>
    </div>
  </transition-stub>
</div>"
`;
